Better registration of write and read-side components in ASP.NET Core applications

jonclare opened the pull request.

jonclare committed:
Start work on improving how components are registered with ASP.NET Core applications

jonclare committed:
Further work on registering write and read-side components in an ASP.NET Core application

jonclare committed:
Remove dependency on EntityFramework from core projections registration library

daniel-smith commented on src/DomainLib.EventPersistence.AspNetCore/AggregateServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,52 @@
+        /// Adds an aggregate repository that uses EventStoreDb to persist events 
+        /// </summary>
+        public static IServiceCollection AddAggregateRepository(this IServiceCollection services,
+                                                                          IConfiguration configuration,
```
nit picky - but formatting here seems a bit out :) Nothing to fix for this PR necessarily though. Would be good to apply some common formatting rules with some tool, at some point.

jonclare replied:
Yep. Shared R# settings?

daniel-smith commented on src/Examples/Shopping.ReadModel/Startup.cs:
```diff
@@ -27,12 +31,18 @@ public void ConfigureServices(IServiceCollection services)
-                                                                       });
+            services.AddReadModel<IDomainEvent>(Configuration,
+                                                options =>
+                                                {
```
Yeah, looks a bit nicer here.

carlosrfernandez commented on src/DomainLib.EventStore.Common.AspNetCore/EventStoreServiceConfigurationExtensions.cs:
```diff
@@ -0,0 +1,42 @@
+    public static class EventStoreServiceConfigurationExtensions
+    {
+        private static readonly object SyncLock = new();
+        private static EventStoreClient _eventStoreClient;
```
Hey Jon,

Can you help me understand why we're storing a static client and doing the lock to create it? 

Do we only ever want one single instance for the lifetime of the application?

jonclare replied:
Yes. We only want one instance, and we can't control how many times a user might call the AddEventStore method, especially if they're using ES for events and snapshots.
If we don't do this, it would potentially overwrite the singleton in the DI container, but anyone who's already resolved it would be using a different ES client instance.
There might be nicer ways of doing this, but it does the job for now

jonclare committed:
Fix DI registration issues

We were previously trying to register things after a service provider had already been build.
Move the registrations earlier in the process so resolution works

carlosrfernandez commented:
LGTM
